### DISCUSSION TITLE
Why is Adsense code (JS) displayed as raw code on my website?

### DISCUSSION THREAD

**Discussion by JetGum** [datenstrom/community#499] [Ask a question] [answered]
Hi,

hope you're all safe and well in these testing times!

A couple of days ago, I noticed that Adsense code embedded in the Markdown files was displaying partially as raw code on the frontent of my website.

Please check: https://www.aussenpool.de. You'll see:
`(adsbygoogle = window.adsbygoogle || []).push({});`

This is actually
`<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>` in the markdown.

Any ideas?

Cheers,

Carsten

**Comment by markseuffert** [comment]
Hi Carsten,

hope you're safe in Germany.

The reason why your old JavaScript code stopped working is that we now use a HTML whitelist. The [Markdown parser](https://github.com/annaesvensson/yellow-markdown) removes unsafe HTML elements/attributes from the content, this feature plays a central role in making secure websites, but nothing stops you from putting them somewhere else. Here are three possible solutions:

1. Put your JavaScript into a layout file, for example `system/layouts/header.html` or `system/layouts/footer.html`. 
2. Write a small extension that does this automatically, for example in event `onParsePageExtra()`.
3. Write your own Markdown parser free from any normalisation, safeguards and restrictions.

Hope this helps and happy coding. 😄

**Comment by JetGum** [comment]
Hi Mark,

Everything's HomeOffice here. Hope situation will improve slowly.

Thanks for your answer. Both solutions are not really viable for me. AdSense code is implemented in different places all over the website (in the *.md files) and I have  different ads on different pages.

I have not yet written an extension to do this. My php knowledge is not profound enough to write an extension to do this.

Could there be another way to include (like shortcode in WordPress) different Is snippets?

Cheers and good health,

Carsten

**Comment by schulle4u** [comment]
You can start with an [example extension](https://github.com/schulle4u/yellow-extension-example). It has a basic shortcode template to help developers integrate html and JS snippets into markdown pages. I'm not an AdSense user, so don't know how difficult it would be to transform the generated code snippets into an extension for Yellow. In theory it should be fairly easy.

**Comment by GiovanniSalmeri** [comment, marked as answer]
I cannot try it (I have never used Adsense), but maybe is it sufficient a tiny extension like this?

```php
<?php
// Adsense extension
// Copyright (c) 2020 Giovanni Salmeri
// This file may be used and distributed under the terms of the public license.

class YellowAdsense {
    const VERSION = "0.8.10";
    const TYPE = "feature";
    public $yellow;         //access to API
    
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
    }
    
    // Handle page content of shortcut
    public function onParseContentShortcut($page, $name, $text, $type) {
        $output = null;
        if ($name=="adsense" && ($type=="block" || $type=="inline")) {
            list($style, $layout, $format, $client, $slot) = $this->yellow->toolbox->getTextArgs($text);
            $output .= "<ins class=\"adsbygoogle\" ";
            $output .= "style=\"{$style}\" ";
            $output .= "data-ad-layout=\"{$layout}\" ";
            $output .= "data-ad-format=\"{$format}\" ";
            $output .= "data-ad-client=\"{$client}\" ";
            $output .= "data-ad-slot=\"{$slot}\"></ins>";
        }
        return $output;
    }
    
    // Handle page extra data
    public function onParsePageExtra($page, $name) {
        $output = null;
        if ($name == "header") {
            $output .= "<script async src=\"//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js\"></script>";
            $output .= "<script defer>(adsbygoogle = window.adsbygoogle || []).push({});</script>";
        }
        return $output;
    }
}
```

Name this file adsense.php, put it in the extension folder and write in your md page something like `[adsense "display:block;text-align:center" in-article fluid ca-pub-5140689302901281 9730069418]`.

**Comment by GiovanniSalmeri** [comment]
Postscriptum: In fact (assuming this extension works) the problem would not be solved by the extension *per se*: a `<script>` tag is currently always deleted in the body of the page, whether written manually or generated by a shortcut (I ask Yellow's core developers for confirmation!). The problem would be solved by moving the `<script>` tag to the page header (as this extension does), with the addition of the attribute `defer` indicating that its execution must be postponed until the whole page is loaded.

**Comment by JetGum** [comment]
Love all your comment and effort. Just tested your solution Giovanni (adsense.php and all) and it seems to work, though it might require some more fiddling :-)

Moving script to the header and defer is a good idea.

I will play with the variables a bit over the weekend. For example, "data-ad-client" is always the same for each particular website, so this could be hardcoded.

Thx so far for your support. Will keep you posted.

**Comment by GiovanniSalmeri** [comment]
Happy to hear that! :-)

Viele Grüße aus Italien!

**Comment by JetGum** [comment]
Hi all,

the extension only seems to work once per page. 
On my homepage ([aussenpool.de](https://www.aussenpool.de/)), only the first Adblock shows. There is another one near the bottom of the page. The code is generated but the ad does not show.

Apart from that, I have reduced the variables to `$style`, `$format` and `$slot`. These variables are different for each ad block I include in the website.

Cheers and lots of health to you all! And stay@home!!!

Carsten

**Comment by GiovanniSalmeri** [comment]
> the extension only seems to work once per page.

Somehow I feared that... Could you try to change this line:

```
$output .= "<script defer>(adsbygoogle = window.adsbygoogle || []).push({});</script>";
```

with this one?

```
$output .= "<script>document.addEventListener('DOMContentLoaded', function() {for (i = document.getElementsByClassName('adsbygoogle').length; i; i--) {(adsbygoogle = window.adsbygoogle || []).push({});}});</script>\n";
```

In this way the script should be executed as many times as there are ads in the page. The JS code seems to me correct, but unfortunately I cannot try this directly... Cheers! :-)

**Comment by JetGum** [comment]
@GiovanniSalmeri 

Nice! Your code works smoothly now, all the ads are showing. Guess we can close now.

Will you put the extension in the repository?

I wish I could write code like that... any tips for a concise php/js tutorial/webinar?

Cheers ans thanks a million!

Carsten
